File: src/main/scala/io/github/oxnz/Ingrid/tx/TxController.scala
package io.github.oxnz.Ingrid.tx

import io.micrometer.core.annotation.Timed
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.http.{HttpStatus, ResponseEntity}
import org.springframework.web.bind.annotation._

import scala.jdk.CollectionConverters._

@RestController
@RequestMapping(Array("tx")) class TxController(val txRecordRepo: TxRecordRepo, val txResultRepo: TxResultRepo, val messageProducer: TxEventProducer) {
  final private val log: Logger = LoggerFactory.getLogger(getClass)

  @GetMapping(Array("records"))
  @Timed def records: List[TxRecord] = {
    txRecordRepo.findAll.asScala.toList
  }

  @GetMapping(Array("results"))
  @Timed def txResults: List[TxResult] = {
    txResultRepo.findAll.asScala.toList
  }

  @PostMapping(Array("record"))
  @Timed(value = "post") def post(@RequestBody request: TxRequest): ResponseEntity[TxResponse] = {
    val entity: TxResponse = process(request)
    new ResponseEntity[TxResponse](entity, HttpStatus.OK)
  }

  private def process(request: TxRequest): TxResponse = {
    var record: TxRecord = new TxRecord(request.cat, request.ref, request.state, request.city)
    record = txRecordRepo.save(record)
    val event: TxEvent = new TxEvent(record.id)
    messageProducer.publish(event)
    new TxResponse(record.id)
  }
}

File: src/main/java/io/github/oxnz/Ingrid/tx/TxController.java
package io.github.oxnz.Ingrid.tx;

import io.github.oxnz.Ingrid.tx.mq.RedisMessageProducer;
import io.github.oxnz.Ingrid.tx.mq.TxEvent;
import io.micrometer.core.annotation.Timed;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("tx")
public class TxController {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private final TxRecordRepo txRecordRepo;
    private final TxResultRepo txResultRepo;
    private final RedisMessageProducer messageProducer;

    public TxController(TxRecordRepo txRecordRepo, TxResultRepo txResultRepo, RedisMessageProducer messageProducer) {
        this.txRecordRepo = txRecordRepo;
        this.txResultRepo = txResultRepo;
        this.messageProducer = messageProducer;
    }

    @GetMapping("records")
    @Timed
    public ResponseEntity<List<TxRecord>> records() {
        List<TxRecord> entities = txRecordRepo.findAll();
        return new ResponseEntity<>(entities, HttpStatus.OK);
    }

    @GetMapping("results")
    @Timed
    public ResponseEntity<List<TxResult>> txResults() {
        List<TxResult> entities = txResultRepo.findAll();
        return new ResponseEntity<>(entities, HttpStatus.OK);
    }

    @PostMapping("record")
    @Timed(value = "post")
    public ResponseEntity<TxResponse> post(@RequestBody TxRequest request) {
        TxResponse entity = process(request);
        return new ResponseEntity<>(entity, HttpStatus.OK);
    }

    private TxResponse process(TxRequest request) {
        TxRecord record = new TxRecord(request.cat, request.ref, request.state, request.city);
        record = txRecordRepo.save(record);
        TxEvent event = new TxEvent(record.id());
        messageProducer.publish(event);
        return new TxResponse(record.id());
    }
}
